test_fms/mpp/test_read_input_nml.F90

```diff
@@ -0,0 +1,88 @@
+!* the terms of the GNU Lesser General Public License as published by
+!* the Free Software Foundation, either version 3 of the License, or (at
+!* your option) any later version.
+!*
+!* FMS is distributed in the hope that it will be useful, but WITHOUT
+!* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+!* FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+!* for more details.
+!*
+!* You should have received a copy of the GNU Lesser General Public
+!* License along with FMS.  If not, see <http://www.gnu.org/licenses/>.
+!***********************************************************************
+
+!> @file
+!! @brief Tests the read_input_nml subroutine in four difference scenarios
+!! @author Colin Gladue
+!! @email gfdl.climate.model.info@noaa.gov
+
+program test_read_input_nml
+
+  use mpp_mod, only : mpp_init, mpp_exit
+  use mpp_mod, only : mpp_error, FATAL, NOTE
+  use mpp_mod, only : read_input_nml, mpp_get_current_pelist_name
+  use mpp_mod, only : input_nml_file, INPUT_STR_LENGTH, get_ascii_file_num_lines
+#include<file_version.h>
+
+character(len=200) :: line !< Storage location of lines read from the input nml
+character(len=128) :: filename !< Name of input nml file to be read
+integer :: stat !< IOSTAT output integer
+integer :: n !< Counts the line number in the file being read
+namelist /test_read_input_nml_nml/ test_numb
+
+open(10, file="test_numb.nml", form="formatted", status="old")
+read(10, nml = test_read_input_nml_nml)
+close(10)
+
+if (test_numb == 1 .or. test_numb == 2 .or. test_numb == 4) then
+  ! Test 1: Tests the subroutine on a valid input nml full of data, 
+  ! with no arguments passed to read_input_nml()
+  ! Test 2: Tests the subroutine on a valid input nml full of data, 
+  ! with a string passed to read_input_nml() in order to read a different nml
+  ! Test 4: Tests the subroutine on a valid empty input nml, 
+  ! with no arguments passed to read_input_nml()
+  if (test_numb == 1 .or. test_numb == 4) then
+    filename = "input.nml"
+    call mpp_init() ! Initialize mpp
+    call read_input_nml()
+  else if (test_numb == 2) then
+    filename = "input_alternative.nml"
+    call mpp_init() ! Initialize mpp
+    call read_input_nml("alternative")
+  end if
+  open(1, file=filename, iostat=stat) ! Open input nml or alternative
+  n = 1
+  do
+    read(1, '(A)', iostat=stat) line
+    if (stat.eq.-1) then
+      exit
+    end if
+    if (input_nml_file(n).ne.line) then
+      call mpp_error(FATAL, "Input nml does not match &
+                                        &the input_nml_file variable")
+    end if
+    n = n + 1
+  end do
+  close(1)
+  call mpp_exit()
+
+else if (test_numb.eq.3) then
+  ! Test 3: Tests with an invalid pelist_name_in pass as an argument. An invalid
+  ! pelist_name_in would be one who's size is greater than local pelist_name
+  call mpp_init ! Initialize mpp        
+  call read_input_nml(mpp_get_current_pelist_name()//"e")
```

Comment:
As a precaution, explicitly define the argument `pelist_name_in=mpp_get_current_pelist_name()//"e"`; some compilers do weird things to unspecified optional arguments.

Reply:
Will do. Thank you.

Reply:
Fixed